### ISSUE TITLE
disk tlv data won't update after studio palette be adjusted to level

### ISSUE THREAD

**Issue by Abel032** [opentoonz/opentoonz#5676]
### Description

disk tlv data won't update when studio palette be adjusted to level

### Steps to Reproduce

https://github.com/user-attachments/assets/440e70b5-17fb-42c8-bef6-005a75dfca57

### Expected Behavior

https://github.com/user-attachments/assets/6ce98c1e-e017-4137-ae18-0b2026a4f1fc

**Comment by RodneyBaker** [comment]
I don't quite follow what is happening in your videos.

If using color styles 0 or 1 be aware that those two color styles are special and reserved... especially color style 0.

**Comment by Abel032** [comment]
> I don't quite follow what is happening in your videos.
> 
> If using color styles 0 or 1 be aware that those two color styles are special and reserved... especially color style 0.
> 
> (I really should watch troubleshooting videos when I'm tired)

Let me explain it.

In first video I adjusted the current tlv's palette to the studio palette "color_key".
Before adjusting:    blue color on canvas refer to color No.2
After adjusting:       blue refer on canvas refer to color No.5
It works well until I reopen the scene, the blue color on canvas turned back to No.2(but not No.5)

In the second video what did different is to paint on canvas so that disk tlv can be update,
so when I reopen the scene,the blue color refer to No.5 normally.

**Comment by Abel032** [comment]
https://github.com/user-attachments/assets/8182f870-3385-44f5-9d70-10b3d27f88c4

**Comment by Abel032** [comment]
The new uploaded videos have sounds, hope them complete the lost info.

**Comment by RodneyBaker** [comment]
Thanks for the updated video.
I am seeing something odd although not quite the same as you.

Two things confuse me about your video (one of which is trivial and the other... a technical mention which I didn't quite follow).

The trivial
Rather than restart to reopen the scene you might just reopen the scene after the Save All via the Recent Files menu.
This unless restarting the program is essential to reproducing the bug.

The technical
You mention in the video that the color blue switches from Style 2 to Style 5 and that throws me off.
Regardless of any change to the palette. lines with a specific style will still refer to that same color style (number) so in this case I would expect your 'Color 2 Blue' writing to turn to white as Style 2 has been replaced by the studio palette's Style 2 which is white.

Now...
Havng said all of this what I think I'm seeing when I attempt to replicate your steps is that the writing goes from original color (blue) to red which is the last color in the Global Palette's color_key palette.  Perhaps you've adjusted the color_key palette so you get a different color (I don't think you opened the Global Palette color_key so I'm not sure what colors are there.  I assume they are the defaults however if you haven't changed them).

So what does all of this mean?
It would seem to suggest the palette file is not getting saved properly.
That or the level itself is getting saved with an incorrect color style index.

I'm leaning toward the latter as the palettes seem to be properly updating.

**Comment by Abel032** [comment]
> This unless restarting the program is essential to reproducing the bug.

Yes, no need to reopen the Opentoonz software, just Save All and reopen the scene from Recent Files Menu.
Restarting the program is not essential.

---

> lines with a specific style will still refer to that same color style (number) so in this case I would expect your 'Color 2 Blue' writing to turn to white as Style 2 has been replaced by the studio palette's Style 2 which is white

I thought that the color blue's Style index should be changed from 2 to 5 because I chose **Adjust Current Level to This Palette**(which should behave like second video shows , without a must to draw on level)
![图片](https://github.com/user-attachments/assets/63f73855-ff88-4ad7-b153-f2bfbad1df59)
What you said might be the content of **Load into Current Level**.

---

> the writing goes from original color (blue) to red which is the last color in the Global Palette's color_key palette

> Perhaps you've adjusted the color_key palette so you get a different color

That's strange because I didn't adjust the color_key.
![图片](https://github.com/user-attachments/assets/cbe629dd-a825-4a9f-aa54-1aa40ee18a2c)

---

> I'm leaning toward the latter as the palettes seem to be properly updating.

I also think so.

**Comment by RodneyBaker** [comment]
Something to keep in mind is that lines (and areas) created with palette styles in toonz raster and vector levels refer to the style index number.  So the level itself is independent of the palette.  It just points to the style that should be used.

This is where the power of these indexed palettes lies as we can swap out palettes and the images automatically update to newly assigned color.  We can do this by transferring palettes into the currrent level's palette such as from the Studio Palette or via FX where we direct Opentoonz to use a different palette.  If we want a line or area to refer to a different style we have to 'paint' it with that style.  This is why Opentoonz has a Brush Tool and... a Paint Brush Tool.  With the Paint Brush Tool we can change what style index number is being pointed to by the level.

This allows us to change colors across entire scenes/productions as we need only adjust the color settings of the indexed style or point the line/area to a new index.

When we swap out palettes we aren't changing the index (as we usually want that to remain the same) but rather what is stored at the source of that index; the palette's style settings.

If/when we perform a method such as copy/pasting a color onto a style or using 'Adjust current level to this palette' we are swapping out the colors (or updating) each style so that it now contains the new settings.  The level still refers to the same style for each of its line/area assignments.

This can get complex as we can reorder/reindex palettes and the change WILL have the index of the line/area automatically follow the style to the new position.  This can be seen when we use Control + Drag to reorder styles in a palette.  The lines/areas won't suddenly change to the color in the old styles position but will adjust the reference to point to the style's new location.

The important thing I suppose is be able to rely on colors not changing unless we want them to change.
There are certainly some nuances to get familiar with.

Apologies if that doesn't all make sense.

**Comment by Abel032** [comment]
So the **'Adjust current level to this palette'** doesn't touch tlv,but only update tpl.
What I did is not expected at first, this method was  designed for swaping out (updating) the several styles in the front of current level's palette.(example: #1~#12)

The difference between **'Load into Current palette'** and **'Adjust current level to this palette'** is that

the former delete styles that out range of the studio palette,but the latter keeps these styles and order them after the last style's index in the studio palette.

I thought I have understood.
